### PR TITLE
[Core] Account for spilled objects when reporting object store memory usage

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
In #21870, feature flag `scheduler_report_pinned_bytes_only` was added to report pinned memory usage as object store memory usage. The feature allows nodes with only secondary copies of objects to report 0 memory usage, so autoscaler can drain and scale down these nodes. However, nodes with 0 pinned memory may host spilled primary copies of objects. Draining these nodes can destroy live objects which slows down or fails a workload.

The fix here is to alway reports a non-zero object store memory usage, when there is 0 pinned memory but live spilled objects. A Python unit test is added that reproduced the issued and verified the fix. The nightly test `autoscaling_shuffle_1tb_1000_partitions` also passed with this fix and `scheduler_report_pinned_bytes_only` enabled.

In the longer term, I suspect we may need to use different properties / semantics for object store memory usage for scheduling, vs whether the object store can be drained.

`scheduler_report_pinned_bytes_only` is enabled again.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_autoscaler_fake_scaledown.py [hunk 1]
```diff
@@ -60,6 +68,114 @@ def recv(self, obj):
...
+        def scaledown_to_one():
+            cpu = ray.cluster_resources().get("CPU", 0)
+            assert cpu > 0, "Scale-down should keep at least 1 node"
+            return cpu == 1
+
+        wait_for_condition(scaledown_to_one, timeout=30)
+
+        # Verify the spilled object still exists, and there is no object in the
+        # plasma store.
+        check_memory([spilled_obj.hex()], num_plasma_objects=0)
```
Comment (stephanie-wang) [inline python/ray/tests/test_autoscaler_fake_scaledown.py:174]: Nice test! Also try deleting the ObjectRef to the `spilled_obj` and check that it scales down to 0 after that?

File: src/ray/raylet/local_object_manager.cc [hunk 2]
```diff
@@ -592,6 +592,18 @@ void LocalObjectManager::RecordMetrics() const {
                                                        "Restored");
 }
 
+int64_t LocalObjectManager::GetPinnedBytes() const {
```
Comment (stephanie-wang) [inline src/ray/raylet/local_object_manager.cc:595]: Add a C++ unit test for this method?

**Review by stephanie-wang — APPROVED** [review]
Nice fix! Left a couple comments about testing.

I also agree with you about decoupling the task scheduler from the autoscaler.

File: python/ray/tests/test_autoscaler_fake_scaledown.py (hunk 1 above)
Comment (mwtian) [reply python/ray/tests/test_autoscaler_fake_scaledown.py:174]: Added.

File: src/ray/raylet/local_object_manager.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/raylet/local_object_manager.cc:595]: Added.

**Review by ericl — COMMENTED** [review]
High level approach sounds good to me. Indeed we should probably split this out into a separate resource like disk_usage in the future, but this will resolve the user issue for now.

Can you also run the nightly test and verify it passes?

**Comment by mwtian** [comment]
Passed a few runs of `autoscaling_shuffle_1tb_1000_partitions` and other autoscaling / shuffle nightly tests. CI failures seem unrelated.

**Comment by ericl** [comment]
Lint error?

**Comment by ericl** [comment]
Hmm guess it's unrelated, merged!

**Pull request merged**